pkg/kube/secrets/manifest.go

```diff
@@ -0,0 +1,241 @@
+func NewManifestPersister(client kubernetes.Interface, namespace string, deploymentName string) ManifestPersister {
+	return &manifestPersisterImpl{
+		client,
+		namespace,
+		deploymentName,
+	}
+}
+
+func (p *manifestPersisterImpl) PersistManifest(manifest manifest.Manifest, sourceDescription string) (*corev1.Secret, error) {
+	currentVersion, err := getGreatestVersion(p)
+	if err != nil {
+		return nil, err
+	}
+
+	version := currentVersion + 1
+
+	secretName, err := secretName(p, version)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := yaml.Marshal(manifest)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.client.CoreV1().Secrets(p.namespace).Create(&corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: p.namespace,
+			Labels: map[string]string{
+				deploymentKeyName: p.deploymentName,
+				versionKeyName:    strconv.Itoa(version),
+				sourceKeyName:     sourceDescription,
+			},
+		},
+		Data: map[string][]byte{
+			manifestKeyName: data,
+		},
+	})
+}
+
+func (p *manifestPersisterImpl) RetrieveVersion(version int) (*corev1.Secret, error) {
+	list, err := p.ListAllVersions()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, secret := range list {
+		ver, err := getVersionFromSecretName(p, secret.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		if ver == version {
+			return &secret, nil
+		}
+	}
+	return nil, fmt.Errorf("unable to find the requested version: %d", version)
+}
+
+func (p *manifestPersisterImpl) RetrieveLatestVersion() (*corev1.Secret, error) {
+	latestVersion, err := getGreatestVersion(p)
+	if err != nil {
+		return nil, err
+	}
+	return p.RetrieveVersion(latestVersion)
+}
+
+func (p *manifestPersisterImpl) ListAllVersions() ([]corev1.Secret, error) {
+	listResp, err := p.client.CoreV1().Secrets(p.namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	result := []corev1.Secret{}
+
+	nameRegex := regexp.MustCompile(fmt.Sprintf(`^deployment-%s-\d+$`, p.deploymentName))
+	for _, secret := range listResp.Items {
+		if nameRegex.MatchString(secret.Name) {
+			result = append(result, secret)
+		}
+	}
+	return result, nil
+}
+
+func (p *manifestPersisterImpl) DeleteManifest() error {
+	list, err := p.ListAllVersions()
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range list {
+		err := p.client.CoreV1().Secrets(p.namespace).Delete(secret.Name, &metav1.DeleteOptions{})
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *manifestPersisterImpl) DecorateManifest(key string, value string) (*corev1.Secret, error) {
+	version, err := getGreatestVersion(p)
+	if err != nil {
+		return nil, err
+	}
+
+	secretName, err := secretName(p, version)
+	if err != nil {
+		return nil, err
+	}
+
+	secret, err := p.client.CoreV1().Secrets(p.namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	labels := secret.GetLabels()
```

Comment:
I read up on labels vs annotations https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/ and it seems we should be using labels for selecting and user facing stuff. An automatic version number used by our service sound like a case for annotations?

Reply:
Yet another interesting point. I have not thought about it much and basically followed what was defined in the tracker. Now that I re-read the Kubernetes documents, I am curios:
> Labels can be used to select objects and to find collections of objects that satisfy certain conditions

With that statement in mind, from the three labels we define, one should be an annotation I would say. The _sources description_ is the only one that is not serving as an identifier. One could argue that you might want to use it to select all secrets that were created by a specific CRD, but without any real use case I would say this sounds like an edge case scenario. The version number and deployment name however are very good to select objects based on a condition.

Reply:
That being said, if we decide on that set of labels and annotations, I would have to rewrite part of the code to actually get a secret based on the labels and not based on the name.

Reply:
@viovanov  any opinion, should we invest time here?

Reply:
I think labels are appropriate for versions in this case.
Keep in mind that the Version here is the iteration of the deployment manifest.
It's useful user-facing information, that can be used to identify things (for a diff in scripts for example).

If I understand this statement correctly:
```
Labels allow for efficient queries and watches and are ideal for use in UIs and CLIs. Non-identifying information should be recorded using annotations.
```

Then I think labels are being used correctly, as they are _identifying_ the deployment manifest, as @HeavyWombat 
 mentioned. I also agree with the exception - `sourceDescription`.
Please make that an annotation, then we can move on.